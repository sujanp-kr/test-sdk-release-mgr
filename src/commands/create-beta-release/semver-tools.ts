import { clean, inc, rcompare as sortSemverNewestToOldest } from "semver";

/**
 * [ v1.0.0, v2.0.1, v2.0.0 ] ==> 2.0.1
 * @param gitTags
 */
export function findNewestVersionFromGitTags(gitTags: string[]): string {
  gitTags = gitTags
    .map((tag) => clean(tag) || "")
    .filter((tagIsNotBlank) => tagIsNotBlank)
    .sort(sortSemverNewestToOldest);

  const newestTag = gitTags[0];
  if (newestTag) {
    return newestTag;
  }

  throw new Error("Unable to find valid semver tag");
}

/**
 * Given newestVersion, returns the next RC version
 *
 * 1.0.0 ==> 1.0.1-rc.0
 *
 * @param newestVersion
 */
export function calculateReleaseCandidateVersion(
  newestVersion: string
): string | null {
  return inc(newestVersion, "prerelease", "rc");
}
