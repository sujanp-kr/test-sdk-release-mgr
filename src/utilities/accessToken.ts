import yargs from "yargs";

/**
 * Loads the github access token from the command line --accessToken option
 * @param commandLineArgs
 */
export function getAccessToken(
  commandLineArgs: string[] = process.argv
): string {
  const cliInput = yargs(commandLineArgs)
    .option({
      accessToken: {
        demandOption: true,
        type: "string",
      },
    })
    .parseSync();

  return cliInput.accessToken;
}
