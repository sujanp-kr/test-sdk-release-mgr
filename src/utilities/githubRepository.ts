import yargs from "yargs";

export type GithubRepo = {
  owner: string;
  repo: string;
};

/**
 * Take --currentRepo github/blah
 *
 * return owner = github ; repo = blah
 *
 * https://docs.github.com/en/actions/learn-github-actions/environment-variables
 *
 * You can pass us env.GITHUB_REPOSITORY
 *
 * @param commandLineArgs
 */
export function getCurrentRepo(
  commandLineArgs: string[] = process.argv
): GithubRepo {
  const cliInput = yargs(commandLineArgs)
    .option({
      currentRepo: {
        demandOption: true,
        type: "string",
      },
    })
    .parseSync();

  const [owner, repo] = cliInput.currentRepo.split("/");

  if (owner && repo) {
    return {
      repo,
      owner,
    };
  }

  throw new Error(
    `Invalid format for repo, expected <owner>/<repo>, got: ${cliInput.currentRepo}`
  );
}
