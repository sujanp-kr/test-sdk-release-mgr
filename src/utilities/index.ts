export * from "./ocktokit";
export * from "./accessToken";
export * from "./pullRequest";
