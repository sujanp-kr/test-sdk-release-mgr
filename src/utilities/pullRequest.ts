import yargs from "yargs";

/**
 * Returns the PR number supplied by --pullRequest from the command line
 * @param commandLineArgs
 */
export function getPullRequest(
  commandLineArgs: string[] = process.argv
): number {
  const cliInput = yargs(commandLineArgs)
    .option({
      pullRequest: {
        demandOption: true,
        type: "number",
      },
    })
    .parseSync();

  return cliInput.pullRequest;
}
