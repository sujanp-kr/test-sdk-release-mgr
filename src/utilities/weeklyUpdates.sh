# This must be run inside from a package.json script

function installJq {
  HOMEBREW_NO_AUTO_UPDATE=1 brew install jq && clear && echo "Successfully installed jq"
}

# Install jq if you don't already have it
which jq || installJq

USER_PATH=".kat/"
rm -rf "$USER_PATH"
mkdir "$USER_PATH"

curl -H "Authorization: token ${ACCESS_TOKEN}" \
  -H "Accept: application/vnd.github.v3+json" \
  -L https://api.github.com/repos/krogertechnology/kat-scenario-poet/releases/latest |
  jq ".assets[0].url" |
  xargs -n 1 curl -L -H "Authorization: token ${ACCESS_TOKEN}" \
    -H "Accept: application/octet-stream" \
    -o "${USER_PATH}scenario-poet-cli-latest.tar"

tar -xf "${USER_PATH}scenario-poet-cli-latest.tar" -C "${USER_PATH}"

SCHEMA_REGISTRY_PATH="${USER_PATH}schema-registry"
rm -rf "$SCHEMA_REGISTRY_PATH"
git clone --quiet --depth 1 "https://${ACCESS_TOKEN}@github.com/krogertechnology/schema-registry.git" "$SCHEMA_REGISTRY_PATH"

"${USER_PATH}scenario-poet-cli/bin/scenario-poet-cli" --root "$SCHEMA_ROOT" --language typescript "$SCHEMA_REGISTRY_PATH" --output "${USER_PATH}OUTPUT_DIR"
rm -rf "${WEEKLY_UPDATES_DESTINATION}/schema"
cp -r "${USER_PATH}OUTPUT_DIR/schema" "${WEEKLY_UPDATES_DESTINATION}"
