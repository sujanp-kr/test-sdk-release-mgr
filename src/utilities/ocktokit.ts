import { Octokit } from "octokit";
// eslint-disable-next-line no-duplicate-imports
import type { Octokit as _OctokitInstance } from "octokit";
import { default as ProxyAgent } from "proxy-agent";
import { getAccessToken } from "./accessToken";

export type OctokitInstance = _OctokitInstance;

export type GithubPullRequest = Awaited<
  ReturnType<OctokitInstance["rest"]["pulls"]["get"]>
>["data"];

export function getOctokitInstance(
  accessToken: string = getAccessToken()
): OctokitInstance {
  return new Octokit({
    auth: accessToken,
    request: {
      // The Kroger space needs this proxy util
      agent: new ProxyAgent(),
    },
  });
}
