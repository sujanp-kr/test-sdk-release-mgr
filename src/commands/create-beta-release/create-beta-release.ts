import { getOctokitInstance, getPullRequest } from "../../utilities";
import { getCurrentRepo } from "../../utilities/githubRepository";
import {
  calculateReleaseCandidateVersion,
  findNewestVersionFromGitTags,
} from "./semver-tools";

export async function createBetaRelease(
  { octokit, repo, pullRequest } = {
    octokit: getOctokitInstance(),
    repo: getCurrentRepo(),
    pullRequest: getPullRequest(),
  }
): Promise<string> {
  const tagRequestFromGithub = await octokit.rest.repos.listTags(repo);

  const newestTag = findNewestVersionFromGitTags(
    tagRequestFromGithub.data.map((github_tag) => github_tag.name)
  );
  console.log("Found newest tag:", newestTag);

  const rcTag = `v${calculateReleaseCandidateVersion(newestTag)}`;

  const prFromGithub = await octokit.rest.pulls.get({
    ...repo,
    pull_number: pullRequest,
  });

  console.log("Got PR:", `#${pullRequest}`, prFromGithub.data?.title);

  // In Github, you must create a tag reference and then release that tag
  const tagRef = await octokit.rest.git.createRef({
    ...repo,
    sha: prFromGithub.data.head.sha,
    ref: `refs/tags/${rcTag}`,
  });

  console.log("Tag ref created:", tagRef?.data?.ref);

  const release = await octokit.rest.repos.createRelease({
    ...repo,
    prerelease: true,
    tag_name: rcTag,
    // Gives the release a title in Github
    name: rcTag,
  });

  console.log("Release created:", release?.data?.html_url);

  await octokit.rest.issues.createComment({
    ...repo,
    // In Github, pull requests are just issues with code
    issue_number: pullRequest,
    body: `Beta release ${rcTag} has been created :tada:\n${release?.data?.html_url}`,
  });

  console.log("Commented on PR");

  return rcTag;
}
