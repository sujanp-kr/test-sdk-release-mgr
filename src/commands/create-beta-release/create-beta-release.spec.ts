import { DeepMockProxy, mockDeep } from "jest-mock-extended";
import { Octokit } from "octokit";
import { GithubPullRequest } from "../../utilities";
import { createBetaRelease as sut } from "./create-beta-release";

describe("createBetaRelease", () => {
  const mocktokit: DeepMockProxy<Octokit> = mockDeep<Octokit>();

  function buildGithubResponse<T>(data: T): {
    data: T;
    headers: { [k: string]: string };
    status: 200;
    url: string;
  } {
    return {
      data,
      headers: {},
      status: 200,
      url: "you wish",
    };
  }

  beforeEach(() => {
    jest.resetAllMocks();
  });

  it("Should return the correct tag", async () => {
    mocktokit.rest.repos.listTags.mockResolvedValueOnce(
      buildGithubResponse([
        {
          name: "v1.0.0",
          commit: {
            sha: "example",
            url: "blah.com/example",
          },
          node_id: "blah-blah",
          tarball_url: "blah",
          zipball_url: "blah",
        },
      ])
    );

    mocktokit.rest.pulls.get.mockResolvedValueOnce(
      buildGithubResponse({
        head: {
          sha: "git-sha-781290389381",
        },
      } as unknown as GithubPullRequest)
    );

    const result = await sut({
      octokit: mocktokit,
      repo: {
        repo: "typescript",
        owner: "Microsoft",
      },
      pullRequest: 7,
    });

    expect(result).toEqual("v1.0.1-rc.0");
  });

  it("Should throw an error if a valid tag is not found", async () => {
    mocktokit.rest.repos.listTags.mockResolvedValueOnce(
      buildGithubResponse([
        {
          name: "invalid-semver-tag",
          commit: {
            sha: "example",
            url: "blah.com/example",
          },
          node_id: "blah-blah",
          tarball_url: "blah",
          zipball_url: "blah",
        },
      ])
    );

    try {
      await sut({
        octokit: mocktokit,
        repo: {
          repo: "MacOS",
          owner: "Apple",
        },
        pullRequest: 15,
      });
    } catch (e: Error | unknown) {
      if (e instanceof Error) {
        expect(e.message).toContain("tag");
        expect(e).toMatchSnapshot();
      } else {
        fail("Should be a real error");
      }
    }
  });

  it("Should make correct calls out to Github", async () => {
    mocktokit.rest.repos.listTags.mockResolvedValueOnce(
      buildGithubResponse([
        {
          name: "v1.0.0",
          commit: {
            sha: "example",
            url: "blah.com/example",
          },
          node_id: "blah-blah",
          tarball_url: "blah",
          zipball_url: "blah",
        },
      ])
    );

    mocktokit.rest.pulls.get.mockResolvedValueOnce(
      buildGithubResponse({
        head: {
          sha: "git-sha-781290389381",
        },
      } as unknown as GithubPullRequest)
    );

    mocktokit.rest.repos.createRelease.mockResolvedValueOnce(
      buildGithubResponse({
        html_url: "github.com/repo/pulls/n",
      }) as never
    );

    await sut({
      octokit: mocktokit,
      repo: {
        repo: "typescript",
        owner: "Microsoft",
      },
      pullRequest: 7,
    });

    expect(mocktokit.rest.git.createRef).toHaveBeenCalledWith({
      repo: "typescript",
      owner: "Microsoft",
      sha: "git-sha-781290389381",
      ref: `refs/tags/v1.0.1-rc.0`,
    });

    expect(mocktokit.rest.repos.createRelease).toHaveBeenCalledWith({
      repo: "typescript",
      owner: "Microsoft",
      prerelease: true,
      tag_name: "v1.0.1-rc.0",
      name: "v1.0.1-rc.0",
    });

    // The release should be created after the ref
    expect(
      mocktokit.rest.repos.createRelease.mock.invocationCallOrder[0]
    ).toBeGreaterThan(mocktokit.rest.git.createRef.mock.invocationCallOrder[0]);

    expect(mocktokit.rest.issues.createComment).toHaveBeenCalledTimes(1);

    expect(mocktokit.rest.issues.createComment.mock.calls[0]).toMatchSnapshot(
      "Comment created"
    );
  });
});
