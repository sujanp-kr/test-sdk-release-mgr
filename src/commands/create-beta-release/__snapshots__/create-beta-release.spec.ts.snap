// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createBetaRelease Should make correct calls out to Github: Comment created 1`] = `
Array [
  Object {
    "body": "Beta release v1.0.1-rc.0 has been created :tada:
github.com/repo/pulls/n",
    "issue_number": 7,
    "owner": "Microsoft",
    "repo": "typescript",
  },
]
`;

exports[`createBetaRelease Should throw an error if a valid tag is not found 1`] = `[Error: Unable to find valid semver tag]`;
